Improvements in AgnosticHelixFinder plus avgEDep and nHitsRatio implementation

ethanmar opened the pull request:
- Added avgEDep and nHitsRatio cut and histograms to AgnosticHelixFinder
- Added avgEDep and nHitsRatio histogram to CalHelixFinder
- Added avgEDep cut to CalHelixFinder
- Added avgEDep cut and histograms to RobustHelixFinder
- Added avgEDep data member to HelixSeed

FNALbuild commented:
Hi @ethanmar,
You have proposed changes to files in these packages:
- RecoDataProducts
- TrkReco
- CalPatRec
- TrkPatRec

which require these tests: build.

@Mu2e/fnalbuild-users, @Mu2e/write have access to CI actions on [main](/Mu2e/Offline/tree/main).





:hourglass: The following tests have been triggered for e74ae788abc022a422d63992de19a6f4bad563b6: build  (Build queue has 20 jobs)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:bangbang: It was not possible to prepare the workspace for this test. This is often caused by merge conflicts - please check and try again.
```
> git diff --check | grep -i conflict
TrkPatRec/fcl/prolog.fcl:569: leftover conflict marker
TrkPatRec/fcl/prolog.fcl:572: leftover conflict marker
TrkPatRec/fcl/prolog.fcl:621: leftover conflict marker
```

| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :x: | e74ae788abc022a422d63992de19a6f4bad563b6 into 684bac4f22a1fa29f6bfd661300432e492dccbdd merge failed |

brownd1978 commented on CalPatRec/src/AgnosticHelixFinder_module.cc:
```diff
@@ -1689,7 +1739,10 @@ namespace mu2e {
       ComboHit hhit(*hit);
       hhit._hphi = _tcHits[i].helixPhi + _tcHits[i].helixPhiCorrection * 2 * M_PI;
       hseed._hhits.push_back(hhit);
+      eDepSum = eDepSum + hhit.energyDep();
```
Please use += for incremental addition.

brownd1978 commented on CalPatRec/src/AgnosticHelixFinder_module.cc:
```diff
@@ -1689,7 +1739,10 @@ namespace mu2e {
+      eDepSum = eDepSum + hhit.energyDep();
+      nComboHits++;
     }
+    float eDepAvg = eDepSum/nComboHits;
```
This calculation doesn't produce the true average, due to the variable number of straw hits in panel hits.  The mathematically correct average is:
size_t nstrawhits(0);
loop ...
eDepSum  += hhit.energyDep()*hhit.nStrawHits();
nstrawhits += hhit.nStrawHits();
end loop ...
eDepAvg = eDepSum/nstrawhits;

brownd1978 commented on CalPatRec/src/CalHelixFinder_module.cc:
```diff
@@ -61,6 +61,7 @@ namespace mu2e {
     _tpart              ((TrkParticle::type)(pset.get<int>("fitparticle"))),
     _fdir               ((TrkFitDirection::FitDirection)(pset.get<int>("fitdirection"))),
     _doSingleOutput     (pset.get<bool>  ("doSingleOutput")),
+    _maxEDepAvg            (pset.get<float>  ("maxEDepAvg")),
```
Use of ParameterSet for module implementation is deprecated. As per Mu2e software policy established several years ago this module must be upgraded to use validated fhicl before this PR can be processed.

brownd1978 commented on CalPatRec/src/CalHelixFinder_module.cc:
```diff
@@ -507,16 +516,22 @@ namespace mu2e {
       //      hhit._hphi = shphi;
       // hhit._flag.merge(StrawHitFlag::resolvedphi);
-
+      eDepSum = eDepSum + hhit.energyDep();
```
Same comment about the EDep average calculation.  The right answer is to implement the average EDep function correctly in ComboHitCollection, and have clients call that.  This needs to be part of this PR.

brownd1978 commented on TrkPatRec/src/RobustHelixFinder_module.cc:
```diff
@@ -535,19 +541,24 @@ namespace mu2e {
 
       ComboHit                hhit(*hit);
       helixData._hseed._hhits.push_back(hhit);
+      eDepSum = eDepSum + hhit.energyDep();
```
Same comment as before for the average edep calculation.

brownd1978 reviewed (changes requested):
This is a very nice development and a good first effort, thanks Ethan. There are some structural issues that must be addressed before this PR can be merged, please work with your supervisor if you have any questions about what is needed or how to implement the required changes.

ethanmar commented:
- Fixed calculation of edepavg in helixfinder modules
- Added edepavg function in ComboHitCollection
- Updated configuration parameters in CalHelixFinder
- Added combineLines function to TZClusterFinder to combine clusters that should not have been separated

gianipez commented on CalPatRec/src/CalHelixFinderAlg.cc:
```diff
@@ -139,48 +139,50 @@ namespace mu2e {
+    _plotall(config.plotall()),
+    _usetarget(config.usetarget()),
+    _maxZTripletSearch(config.maxZTripletSearch()),
+    _bz(config.bz()),
```
this needs to be removed

gianipez replied:
I'm referring to the _bz(....). That parameter is set in the BeginRun, so it doesn't need to be initialized here.

gianipez commented:
@brownd1978 could you take a look when you have time? I think Ethan addressed all the comments you made if I'm not mistaken.

ethanmar commented:
Addressed Giani's comments

brownd1978 reviewed (changes requested):
Thanks for updating CalHelixFinder to use validated fcl. I note there are still outstanding issues in CalPatRec. Those can be addressed after this PR is merged, but before any other new developments.
As discussed in the meeting, using a class to make the avgedep calculation and cut will make this code more maintainable and reliable. I suggest that class be constructed from an art::table, and that it operate on a helix seed or ComboHit collection, returning a bool. That way all the helix finders will use the same code, whose parameters (currently max <EDep>) can be  configured in a single prolog stanza for all the helix finders, presenting a single point of maintenance. This design is also open to extension, should we figure out other helix-level cuts we want to make. If these instructions are unclear please contact me.

brownd1978 commented:
one other issue: this PR updates apr, tpr, and cpr to use maxavgedep, but not mpr. After you've made the above changes it should be trivial to add that cut to mpr as well, please do so for consistency.

gianipez commented:
> one other issue: this PR updates apr, tpr, and cpr to use maxavgedep, but not mpr. After you've made the above changes it should be trivial to add that cut to mpr as well, please do so for consistency.

This is fine by me, but the `mpr` developers should take the time to check if the same cut value is good or not as this might depend on the purity of the helix candidates.

ethanmar commented:
- Addressed @brownd1978 's comments by implementing the EDepAvg function from ComboHitCollection into all HelixFinder modules, including RobustMultiHelixFinder
- Additionally, I centralized the EDepAvg cut by creating a HelixFinderParams stanza in TrkHitReco/fcl/prolog.fcl which sets the cut. This is then fed into each HelixFinder module in the corresponding prologs

brownd1978 commented on TrkHitReco/fcl/prolog.fcl:
```diff
@@ -147,6 +147,14 @@ TrkHitReco : { @table::TrkHitReco
    }
 }
+
+TrkHitReco : { @table::TrkHitReco
```
My apologies I must have mis-spoke, this stanza belongs in TrkReco, not TrkHitReco.

brownd1978 commented on RecoDataProducts/src/ComboHit.cc:
```diff
@@ -221,6 +221,19 @@ namespace mu2e {
     return retval;
   }
 
+  float ComboHitCollection::eDepAvg(ComboHitCollection const& ccol) {
```
I don't understand why a ComboHitCollection is provided as argument since this is a member function of ComboHitCollection. This function can just act on its own object, with no argument. Also the function should be declared 'const'.

brownd1978 commented on CalPatRec/src/CalHelixFinder_module.cc:
```diff
@@ -77,18 +78,21 @@ namespace mu2e {
 //-----------------------------------------------------------------------------
       _helTraj          = 0;
-      _timeOffsets      = new fhicl::ParameterSet(pset.get<fhicl::ParameterSet>("TimeOffsets",fhicl::ParameterSet()));
+      //_timeOffsets      = new fhicl::ParameterSet(pset.get<fhicl::ParameterSet>("TimeOffsets",fhicl::ParameterSet()));
```
please remove commented lines. Git will keep track of the history.

brownd1978 commented on CalPatRec/src/CalHelixFinderDiag_tool.cc:
```diff
@@ -39,10 +41,13 @@ namespace mu2e {
+  CalHelixFinderDiag::CalHelixFinderDiag(const fhicl::Table<mu2e::CalHelixFinderTypes::Config>& config) {
 }
 
+// CalHelixFinderDiag::CalHelixFinderDiag(const fhicl::ParameterSet& PSet) {
```
please remove commented lines.

brownd1978 commented on CalPatRec/src/CalHelixFinderDiag_tool.cc:
```diff
@@ -17,6 +17,7 @@
 namespace mu2e {
 
   using namespace CalHelixFinderTypes;
+  //using CalHelixFinderTypes::Config;
```
please remove commented lines.

brownd1978 commented on CalPatRec/src/CalHelixFinderAlg.cc:
```diff
@@ -2092,8 +2094,19 @@ namespace mu2e {
+    }
+
+    if (nHitsRatio > _maxNHitsRatio) goto PATTERN_RECOGNITION_END;
     //2017-10-05 Gianipez added the following line to make some tests
```
Is this comment still relevant? Git will keep track of changes and who made them, but it sounds like a test may have been left in?

gianipez replied:
What was tested became permanent. Please don't remove the compiled code.

kutschke replied:
I agree with Dave that a comment like this is not helpful.  If it's important to explain why the test is there then explain it.  If not, the test stands on its own.

gianipez replied:
I agree with removing the comment. I said please don't remove the compiled code, aka, line 2111.

brownd1978 commented on CalPatRec/src/CalHelixFinderAlg.cc:
```diff
@@ -2092,8 +2094,19 @@ namespace mu2e {
+
+    if (nHitsRatio > _maxNHitsRatio) goto PATTERN_RECOGNITION_END;
     //2017-10-05 Gianipez added the following line to make some tests
     if (Helix._szphi.qn() == 0.)                                 goto  PATTERN_RECOGNITION_END;
```
goto is a deprecated programming construct. Please replace these with nest if blocks.

gianipez replied:
I think that this comment is arguable and C++ supports the `goto` command. I understand that some people might think that it is not simple to read, but its use is not wrong and people should be more open to different code styles.

kutschke replied:
I agree with Dave that these not appropriate uses of goto.  Modern flow control constructs exist so that the compiler can better understand the intent of the programmer and produce better optimized code and better check for errors.

gianipez replied:
I believe this is a legit use of `goto` as it makes crystal clear w/o the need for any comments where the condition control ends. I debugged this module many times with Valgrind and I never faced any issue with these `goto` commands. I understand that they can be replaced with nested ifs, but IMHO in this way the code is clearer. So unless there are proofs of inefficiencies or bugs, these comments are speculations driven by a style choice.

brownd1978 replied:
As discussed outside this thread we will grandfather the use of goto in this class for this PR. The downsides of goto programming style are real but should be discussed in a separate forum.

brownd1978 commented on CalPatRec/inc/CalHelixFinder_module.hh:
```diff
@@ -30,6 +30,7 @@
 #include "BTrk/TrkBase/HelixParams.hh"
 #include "BTrk/TrkBase/TrkPoca.hh"
 #include "BTrk/TrkBase/TrkMomCalculator.hh"
+// #include "BTrk/TrkBase/TrkParticle.hh"
```
please remove commented lines

brownd1978 reviewed (changes requested):
Thanks for standardizing the maxavgedep cut. Unfortunately I misguided you on where to put the stanza, it should be in TrkReco, not TrkHitReco.
There are a number of commented-out code lines that need to be removed.
There are some comments about 'test codes' that need to be reviewed and acted if necessary.
The use of goto is deprecated. I understand that is legacy to this module but this PR extends its use. Please replace with nested if blocks.
After these changes I think this is ready to go.

brownd1978 commented:
After discussing with Rob we've agreed the use of goto should be grandfathered in to this PR as it is legacy code. This doesn't change the fact that it's a deprecated coding paradigm for many reasons (code optimization, readability, debugging, ...). The comments still need to be cleaned up. Note that git itself keeps track of who makes changes when (see 'git blame') so this kind of comments is mostly unnecessary.

ethanmar commented:
- Removed unnecessary comments
- Fixed EDepAvg function in ComboHitCollection so that it no longer takes the calling ComboHitCollection as an argument
- Moved standardization of maxedepavg cut to TrkReco
- Complementary changes to mu2e_trig_config can be found at following PR: [https://github.com/Mu2e/mu2e_trig_config/pull/35](url)

brownd1978 reviewed (approved):
Thanks for your patience processing this PR Ethan. The result looks good to me.

Please note that migrating to APR in our main sequences requires coordinated downstream changes in Production as well as mu2e_trig_config. Trigger algorithm changes also need to be reviewed in a different forum from this one, where we just deal with code quality issues. We want to use this as a test case for developing a procedure for making trigger algorithm changes in a consistent way for both Online and Offline, please contact Sarah if there are questions.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 0ff1878b7bf5a076b4708a0c1c5aaef48d83468b: build  (Build queue is empty)

FNALbuild commented:
:umbrella: The build tests failed for 0ff1878b7bf5a076b4708a0c1c5aaef48d83468b.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 0ff1878b7bf5a076b4708a0c1c5aaef48d83468b at e8fecef2d6946b1427340b595c2fed5e16c645ec |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/scons.log). Build time: 08 min 27 sec |
| ceSimReco | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/ceSimReco.log) Return Code 9. |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/cosmicOffSpill.log) Return Code 9. |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/ceSteps.log)  |
| ceDigi | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/ceDigi.log) Return Code 9. |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/muDauSteps.log)  |
| ceMix | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/ceMix.log) Return Code 9. |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :large_orange_diamond: | [TODO (5) FIXME (31) in 20 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 3405 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 0ff1878b7bf5a076b4708a0c1c5aaef48d83468b after being merged into the base branch at e8fecef2d6946b1427340b595c2fed5e16c645ec.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2210/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

ethanmar commented:
Updated CalPatRec/fcl/prolog.fcl to match changes in reco modules

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 8408702088dfaee280d0a72f920f82c99203901a: build  (Build queue has 1 jobs)

FNALbuild commented:
:umbrella: The build tests failed for 8408702088dfaee280d0a72f920f82c99203901a.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 8408702088dfaee280d0a72f920f82c99203901a at e8fecef2d6946b1427340b595c2fed5e16c645ec |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/scons.log). Build time: 04 min 13 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/cosmicOffSpill.log) Return Code 9. |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/ceSteps.log)  |
| ceDigi | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/ceDigi.log) Return Code 9. |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/muDauSteps.log)  |
| ceMix | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/ceMix.log) Return Code 9. |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :large_orange_diamond: | [TODO (5) FIXME (31) in 20 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 3405 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 8408702088dfaee280d0a72f920f82c99203901a after being merged into the base branch at e8fecef2d6946b1427340b595c2fed5e16c645ec.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2212/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

ethanmar commented:
@brownd1978 do you need run the build test with my mu2e_trig_config PR? It seems like that would fix what's going wrong with these last three tests. If you do, it can be found here: [https://github.com/Mu2e/mu2e_trig_config/pull/35](url)

kutschke commented:
@ethanmar - you can run the build with tests yourself.  I forget if "test with" works with mu2e_trig_config but you can try. See https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild

kutschke commented:
I am not sure that this will work but I am trying it now:

@FNALbuild run build test with Mu2e/mu2e_trig_config#35

FNALbuild commented:
:sunny: The build tests passed at 8408702088dfaee280d0a72f920f82c99203901a.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :x: | Failed to retrieve request base branch commit sha for Mu2e/mu2e_trig_config#35, default is to merge into main |
| test with | :white_check_mark: | Included Mu2e/mu2e_trig_config#35 @ 5a555ee20ab3155581317e34137afc404e1e75b6 by merge |
| merge | :white_check_mark: | Merged 8408702088dfaee280d0a72f920f82c99203901a at e8fecef2d6946b1427340b595c2fed5e16c645ec |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/scons.log). Build time: 04 min 13 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :large_orange_diamond: | [TODO (5) FIXME (31) in 20 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 3405 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 8408702088dfaee280d0a72f920f82c99203901a after being merged into the base branch at e8fecef2d6946b1427340b595c2fed5e16c645ec.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2213/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

ethanmar commented:
Sorry I had to hop on a call. Looks like it worked? Not sure what the first failure means

kutschke commented:
> Sorry I had to hop on a call. Looks like it worked? Not sure what the first failure means

The message is informational, not a hard failure.  It comes from this line: https://github.com/Mu2e/codetools/blob/ee9bae97cc24ec43e7872b95b2e455f66f1c3607/bin/github/github_common.sh#L78

A little later on you see the code that generates the "test with" line in the report
https://github.com/Mu2e/codetools/blob/ee9bae97cc24ec43e7872b95b2e455f66f1c3607/bin/github/github_common.sh#L100

In my opinion this PR has passed the CI tests.  We need to merge this PR and the one in mu2e_trig_config at the same time.

brownd1978 commented:
mu2e_trig_config PR 35 adds new trigger lines in addition to
fixing compatibility with Ethan's code changes.  The new trigger lines will
change the output streams currently selected in Production, which is an
unwanted side effect that will make it impossible to understand the
pure algorithmic impact in the nightly validation plots. To make this
behave just as an algorithm update we need to make a coordinated update to
Production before this PR (and 35) can be merged.

I'd like to have a general discussion about managing trigger lines. We need
to develop and test, but we also need stability for production, and the
ability to validate the effect of individual changes. Our current
procedures and conventions couple these all together.
Fixing this is beyond the scope of this PR but unfortunately it's coupled
to it.

On Tue, Jul 30, 2024 at 2:04 PM Rob Kutschke ***@***.***>
wrote:

> Sorry I had to hop on a call. Looks like it worked? Not sure what the
> first failure means
>
> The message is informational, not a hard failure. It comes from this line:
> https://github.com/Mu2e/codetools/blob/ee9bae97cc24ec43e7872b95b2e455f66f1c3607/bin/github/github_common.sh#L78
>
> A little later on you see the code that generates the "test with" line in
> the report
>
> https://github.com/Mu2e/codetools/blob/ee9bae97cc24ec43e7872b95b2e455f66f1c3607/bin/github/github_common.sh#L100
>
> In my opinion this PR has passed the CI tests. We need to merge this PR
> and the one in mu2e_trig_config at the same time.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/1289#issuecomment-2259206462>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57636GAW63MJFCCVQFTZO75VLAVCNFSM6AAAAABJWYFV7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENJZGIYDMNBWGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

ethanmar commented:
Hi @brownd1978, I'm not very familiar with the Production repository. Are there certain changes to certain files you need me to make there?

gianipez commented:
I think there is no arm in merging both PRs as it was also previously done when the ‘mpr_’ sequence was added. The MC production or reprocessing campaign should use a well defined tagged version of mu2e_trig_config with a physMenu customized for the purpose of the campaign.For what concerns the nightly validation, hopefully soon we will have integrated a new analyzer module that will provide more clean info about the online reco alg behavior, so, for this transition phase, I would just use patience to interpret the results (as it was done when the mpr sequence was added).GianiSent from my iPhoneOn Jul 30, 2024, at 11:55 PM, Ethan Martinez ***@***.***> wrote:﻿
Hi @brownd1978, I'm not very familiar with the Production repository. Are there certain changes to certain files you need me to make there?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

FNALbuild commented:
:memo: The HEAD of `main` has changed to b5d3931a11597b4f06aa12c5c29d3044b4b6a735. Tests are now out of date.

brownd1978 commented:
Hi Giani,
The point of the nightly is to allow adiabatic testing of new code so we
can understand the impact of changes. One principle of that is not putting
multiple major changes in at the same time. Rob and I have followed this
principle for many years. Ethan’s improvement to track finding is an
important change and I would very much like to see its impact, but by
integrating that change with the addition of new trigger lines in
Production PR 35 it will be impossible to understand its effect by itself.
Patience won’t help: we get one shot to see these differences as the code
base is constantly evolving.

There are two possible fixes. The easiest is to split PR 35 into 2 parts,
one with just the updates needed for APR the other just the new lines. We
can then test them in successive nights. This is the recommended procedure,
ie we ask people not to mix unrelated changes in the same PR.
The other solution is to make the scripts run by the nightly exclude the
new trigger lines, so we maintain the same baseline content. That’s more
stable but will prevent the nightly from being able to evaluate the new
lines, at least for now.

In general I feel we need better communication and coordination of trigger
development work. As a shared online and offline resource, and as the
selector of the physics content of the experiments data, its development
affects many parties. Let’s work together with Sarah to figure out how we
can make efficient progress while respecting everyone’s needs.
Dave

David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720


On Wed, Jul 31, 2024 at 04:37 Gianantonio Pezzullo ***@***.***>
wrote:

> I think there is no arm in merging both PRs as it was also previously done
> when the ‘mpr_’ sequence was added. The MC production or reprocessing
> campaign should use a well defined tagged version of mu2e_trig_config with
> a physMenu customized for the purpose of the campaign.For what concerns the
> nightly validation, hopefully soon we will have integrated a new analyzer
> module that will provide more clean info about the online reco alg
> behavior, so, for this transition phase, I would just use patience to
> interpret the results (as it was done when the mpr sequence was
> added).GianiSent from my iPhoneOn Jul 30, 2024, at 11:55 PM, Ethan Martinez
> ***@***.***> wrote:﻿
> Hi @brownd1978, I'm not very familiar with the Production repository. Are
> there certain changes to certain files you need me to make there?
>
> —Reply to this email directly, view it on GitHub, or unsubscribe.You are
> receiving this because you commented.Message ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/1289#issuecomment-2260314304>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57ZFMAMTRY7DUB2CVYTZPDD7RAVCNFSM6AAAAABJWYFV7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENRQGMYTIMZQGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

matthewstortini commented:
Can we just take out the new lines for now, as Dave asked? This PR has been in progress for over a month. I would like it to go through since I am making enough changes in Ethan's Offline area (in his VNC) to warrant another PR once this one is complete. Ethan is gone the rest of the week, so shall I remove the new highP trigger lines in Ethan's area and re-commit, @gianipez?

Sorry for inserting myself into a conversation. I am only doing so because new changes I'd like to commit (which address issue #1270 and some other improvements in TZClusterFinder) are being developed in one of Ethan's areas. We are developing there as a way of training Ethan. Otherwise I'd just make my own PR.

brownd1978 commented:
I am testing a patch to Production to insulate validation from the new
trigger lines. There is also a backlog of other PRs that need merging, once
that's clear we'll merge these in a block.

On Thu, Aug 1, 2024 at 12:26 PM matthewstortini ***@***.***>
wrote:

> Can we just take out the new lines for now, as Dave asked? This PR has
> been in progress for over a month. I would like it to go through since I am
> making enough changes in Ethan's Offline area (in his VNC) to warrant
> another PR once this one is complete. Ethan is gone the rest of the week,
> so shall I remove the new highP trigger lines in Ethan's area and
> re-commit, @gianipez <https://github.com/gianipez>?
>
> Sorry for inserting myself into a conversation. I am only doing so because
> new changes I'd like to commit (which address issue #1270
> <https://github.com/Mu2e/Offline/issues/1270> and some other improvements
> in TZClusterFinder) are being developed in one of Ethan's areas. We are
> developing there as a way of training Ethan. Otherwise I'd just make my own
> PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/1289#issuecomment-2263808770>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57YKHTXEMMERP5OHZUTZPKDVDAVCNFSM6AAAAABJWYFV7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENRTHAYDQNZXGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

ethanmar commented:
Sorry for my absence at the end of last week. I've just commented out the new highP trigger lines in the mu2e_trig_config PR 35, which will be uncommented in a separate PR. Is this ready to be merged now?
